### PR TITLE
Ensure all entries are cleared from cache on runtime change

### LINKED ISSUE #18917: Context declared in _app undefined during SSR in dev
# Bug report

## Describe the bug

Running into basically this issue #4194 here _sometimes_, only in dev.

It's a fresh app, not much happening as of yet. A single context provider wrapping `Component` in app, and a index page using the context through a custom hook.
The contexts initial value is null, but its filled through the provider. The custom hook throws on empty context value. 

## To Reproduce

Repo: https://github.com/ljosberinn/AuctionCraftSniper/blob/518d537557/src/pages/_app.tsx

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Check that repo out (branch `next`)
2. `yarn dev`
3. click the link in the console to localhost:3000
4. first render will work
5. F5, crashes

## Expected behavior

Context should not be empty.

## Screenshots

_app:
image

AuthContextProvider rendering _with a value present_
image

index calling hook:
image

Breakpoint where error is thrown:
image

Call Stack in VSCode Debugger:
image
Call Stack in Browser:
image
Call Stack in Console:
image

## System information

- OS: Win 10
- Browser (if applies) Chrome
- Version of Next.js: latest
- Version of Node.js: 15.0.1

### PR DESCRIPTION
**Pull request by ijjk**
When leveraging webpack 5, a shared runtime chunk is used on the server between entries which is cleared in development when a change occurs. Currently not all entries are cleared when clearing the runtime chunk causing some entries to use a stale runtime chunk which breaks cases like using `react` context in multiple entries to break. This fixes it by ensuring all entries are cleared when the runtime chunk is cleared and adds a test to prevent regression. 

Fixes: https://github.com/vercel/next.js/issues/18917

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| buildDuration | 11s | 10.7s | -374ms |
| nodeModulesSize | 83 MB | 83 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.132 | 2.183 | ⚠️ +0.05 |
| / avg req/sec | 1172.57 | 1145.34 | ⚠️ -27.23 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.542 | 1.566 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 1621.11 | 1596.65 | ⚠️ -24.46 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| 677f882d2ed8..2428.js gzip | 12.9 kB | 12.9 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-212f85f..f23b.js gzip | 6.59 kB | 6.59 kB | ✓ |
| webpack-7193..1446.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.2 kB | 59.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| polyfills-67..b7d1.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _app-6220e08..9a40.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-4b0b5..2c91.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-5f309a..7282.js gzip | 887 B | 887 B | ✓ |
| index-57f580..c562.js gzip | 227 B | 227 B | ✓ |
| link-b862cd7..dba8.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bd82.js gzip | 303 B | 303 B | ✓ |
| withRouter-2..e384.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.08 kB | 8.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| index.html gzip | 617 B | 617 B | ✓ |
| link.html gzip | 622 B | 622 B | ✓ |
| withRouter.html gzip | 610 B | 610 B | ✓ |
| Overall change | 1.85 kB | 1.85 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| buildDuration | 12.9s | 12.7s | -176ms |
| nodeModulesSize | 83 MB | 83 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| 677f882d2ed8..2428.js gzip | 12.9 kB | 12.9 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-212f85f..f23b.js gzip | 6.59 kB | 6.59 kB | ✓ |
| webpack-7193..1446.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.2 kB | 59.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| polyfills-67..b7d1.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _app-6220e08..9a40.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-4b0b5..2c91.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-5f309a..7282.js gzip | 887 B | 887 B | ✓ |
| index-57f580..c562.js gzip | 227 B | 227 B | ✓ |
| link-b862cd7..dba8.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bd82.js gzip | 303 B | 303 B | ✓ |
| withRouter-2..e384.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.08 kB | 8.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.17 MB | 5.17 MB | ✓ |

</details>

</details>
Commit: c306835290ca02686631bb694b5dbddaa8916bbe

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| buildDuration | 11.3s | 12.1s | ⚠️ +858ms |
| nodeModulesSize | 83 MB | 83 MB | ⚠️ +836 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.329 | 2.468 | ⚠️ +0.14 |
| / avg req/sec | 1073.36 | 1012.87 | ⚠️ -60.49 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.701 | 1.9 | ⚠️ +0.2 |
| /error-in-render avg req/sec | 1469.35 | 1316.13 | ⚠️ -153.22 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| 677f882d2ed8..2428.js gzip | 12.9 kB | 12.9 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-212f85f..f23b.js gzip | 6.59 kB | 6.59 kB | ✓ |
| webpack-7193..1446.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.2 kB | 59.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| polyfills-67..b7d1.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _app-6220e08..9a40.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-4b0b5..2c91.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-5f309a..7282.js gzip | 887 B | 887 B | ✓ |
| index-57f580..c562.js gzip | 227 B | 227 B | ✓ |
| link-b862cd7..dba8.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bd82.js gzip | 303 B | 303 B | ✓ |
| withRouter-2..e384.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.08 kB | 8.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| index.html gzip | 617 B | 617 B | ✓ |
| link.html gzip | 622 B | 622 B | ✓ |
| withRouter.html gzip | 610 B | 610 B | ✓ |
| Overall change | 1.85 kB | 1.85 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| buildDuration | 14.1s | 14.2s | ⚠️ +163ms |
| nodeModulesSize | 83 MB | 83 MB | ⚠️ +836 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| 677f882d2ed8..2428.js gzip | 12.9 kB | 12.9 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-212f85f..f23b.js gzip | 6.59 kB | 6.59 kB | ✓ |
| webpack-7193..1446.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.2 kB | 59.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| polyfills-67..b7d1.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _app-6220e08..9a40.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-4b0b5..2c91.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-5f309a..7282.js gzip | 887 B | 887 B | ✓ |
| index-57f580..c562.js gzip | 227 B | 227 B | ✓ |
| link-b862cd7..dba8.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bd82.js gzip | 303 B | 303 B | ✓ |
| withRouter-2..e384.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.08 kB | 8.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.17 MB | 5.17 MB | ✓ |

</details>

</details>
Commit: 9fe4d0e3bfbd519f9dc3e51e67218978531e71c3

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| buildDuration | 11.9s | 11.5s | -345ms |
| nodeModulesSize | 83 MB | 83 MB | ⚠️ +819 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.364 | 2.333 | -0.03 |
| / avg req/sec | 1057.36 | 1071.76 | +14.4 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.67 | 1.616 | -0.05 |
| /error-in-render avg req/sec | 1496.62 | 1547.19 | +50.57 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| 677f882d2ed8..2428.js gzip | 12.9 kB | 12.9 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-212f85f..f23b.js gzip | 6.59 kB | 6.59 kB | ✓ |
| webpack-7193..1446.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.2 kB | 59.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| polyfills-67..b7d1.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _app-6220e08..9a40.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-4b0b5..2c91.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-5f309a..7282.js gzip | 887 B | 887 B | ✓ |
| index-57f580..c562.js gzip | 227 B | 227 B | ✓ |
| link-b862cd7..dba8.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bd82.js gzip | 303 B | 303 B | ✓ |
| withRouter-2..e384.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.08 kB | 8.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| index.html gzip | 617 B | 617 B | ✓ |
| link.html gzip | 622 B | 622 B | ✓ |
| withRouter.html gzip | 610 B | 610 B | ✓ |
| Overall change | 1.85 kB | 1.85 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| buildDuration | 13.5s | 13.7s | ⚠️ +244ms |
| nodeModulesSize | 83 MB | 83 MB | ⚠️ +819 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| 677f882d2ed8..2428.js gzip | 12.9 kB | 12.9 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-212f85f..f23b.js gzip | 6.59 kB | 6.59 kB | ✓ |
| webpack-7193..1446.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.2 kB | 59.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| polyfills-67..b7d1.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _app-6220e08..9a40.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-4b0b5..2c91.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-5f309a..7282.js gzip | 887 B | 887 B | ✓ |
| index-57f580..c562.js gzip | 227 B | 227 B | ✓ |
| link-b862cd7..dba8.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bd82.js gzip | 303 B | 303 B | ✓ |
| withRouter-2..e384.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.08 kB | 8.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js wp5-runtime-cache | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.17 MB | 5.17 MB | ✓ |

</details>

</details>
Commit: da5e7e41f4d099dd8b45de7489cfafa94a2d6ef9

**Review by Timer — APPROVED** [review]

File: test/integration/ssr-ctx/pages/index.js [hunk 1]
```diff
@@ -1,14 +1,15 @@
 import React from 'react'
-
-const Idk = React.createContext(null)
+import { useIdk } from '../context'
 
 const Page = () => {
+  const idk = useIdk()
+
+  console.log(idk)
```
Comment (PepijnSenders) [inline test/integration/ssr-ctx/pages/index.js:7]: You left a console.log here 😉

Comment (ijjk) [reply test/integration/ssr-ctx/pages/index.js:7]: This was on purpose, it's to see when the test is failing while running

**Pull request merged**